File: 0145_easy_hot_binary-tree-postorder-traversal/solution.java
import java.util.ArrayList;
import java.util.List;
import java.util.Deque;
import java.util.LinkedList;

/**
 * Definition for a binary tree node.
 */

 class TreeNode {
     int val;
     TreeNode left;
     TreeNode right;
     TreeNode() {}
     TreeNode(int val) { this.val = val; }
     TreeNode(int val, TreeNode left, TreeNode right) {
         this.val = val;
         this.left = left;
         this.right = right;
     }
 }

class Solution {
    public List<Integer> postorderTraversal2(TreeNode root) {
        List<Integer> result = new ArrayList<>();
        Deque<TreeNode> stack = new LinkedList<>();

        TreeNode prevAccess = null;
        while(root != null || !stack.isEmpty()) {
            while(root != null) {
                stack.push(root);
                root=root.left;
            }

            root=stack.pop();
            if(root.right == null || root.right == prevAccess) {
                result.add(root.val);
                prevAccess = root;
                root = null;
            } else {
                stack.push(root);
                root=root.right;
            }
        }
        return result;
    }

    public List<Integer> postorderTraversal(TreeNode root) {
        List<Integer> result = new ArrayList<>();

        if(root != null) {
            result.addAll(postorderTraversal(root.left));
            result.addAll(postorderTraversal(root.right));
            result.add(root.val);
        }
        
        return result;
    }
}

File: 0145_easy_hot_binary-tree-postorder-traversal/solution.py
# Definition for a binary tree node.
from collections import deque
from typing import List, Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

        
class Solution:
    def postorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        if not root:
            return []

        return self.postorderTraversal(root.left) + self.postorderTraversal(root.right) + [root.val]
    
    def postorderTraversal2(self, root: Optional[TreeNode]) -> List[int]:
        result = []
        stack = deque()

        prev_access = None

        while root or stack:
            while root:
                stack.append(root)
                root = root.left
            
            root = stack.pop()

            if root.right is None or root.right == prev_access:
                result.append(root.val)
                prev_access = root
                root = None
            else:
                stack.append(root)
                root = root.right
        
        return result